### ISSUE TITLE
modifying config with submodules

### ISSUE THREAD

**Issue by themolecule** [jrburke/requirejs#1485]
So, it would seem that modifying the require.config() is not possible after main.js has happened.

I find documentation online that says otherwise, but I can't figure out what is wrong with my setup.

I can provide code, and will, but posting the concept first to make sure it's solid:

My index.html refers to data-main=js/main, and that loads fine.

Then, in order to provide a progress bar and other error checking, main.js defines a require.config to set up basic paths, and intercept require.load, etc.  Ultimately, though, main.js uses a localRequire to load the rest of the libraries and continue on its way.

From there, it's all dynamic.  Depending on what is the state of things and what is loaded by which user, and the state of the resources that are being loaded, a lot of things may happen.

What I want to have happen is that a library is loaded, and at its option may require() something else, but from an entirely different place on the filesystem.  I was thinking that a simple prefix string, like 'MEDIA/' could be added to the require.config.paths object, and 'MEDIA/' would resolve to whatever I want, and all requests below that would be relative to the parent caller.

What I'm finding is that this dynamism doesn't happen... 'require' the global object is not the same as a 'localRequire', and require.config is a little fragile and can only be set once.  Subsequent requests, IE dependencies needed by the dynamically require()d library, refer to the path set by main.js, and don't inherit the base path of their parent, and setting a prefix path like 'MEDIA/' doesn't resolve this.

Am I:
1) hitting on a limitation of the config system
2) doing it the wrong way, ie, by using a localRequire to build the whole site, instead of a straight define()
3) revealing a bug of some kind where newly added config settings don't cause their children to inherit things correctly.

Another idea: if dynamically added require() requests use a protocol, like "http://", does that reset the context for that group of dependencies?  This I will test next.

thanks!
Chris

**Comment by jrburke** [comment]
It is hard to know the exact problem just going on a description, but here are some things that come to mind:

It is best to use define() when defining a piece of functionality, a module. Only use `require([])` when wanting to trigger a loading of functionality.

requirejs.config() can be called multiple times, but .config() is not on a local require. So this would fail:

```javascript
define(function(require) {
  //fails, does not exist
  require.config();
});
```

This works, since it uses the global requirejs:

```javascript
define(function(require) {
  //Works
  requirejs.config();
});
```

However, multiple config calls only work after the config() call has been done. So in the above examples, if those modules have dependencies, the config() call would only be called after those dependencies are resolved and executed, and the config() change only takes place after that point.

Note that .config() modifies the global requirejs config, and so the config objects need to operate on module IDs that are the global, finally resolved IDs, do not work with relative module IDs.

Hopefully that helps. Closing as a discussion issue, but feel free to continue discussion here and we can reopen as necessary.

**Comment by themolecule** [comment]
Thanks! Requests using http:// allowed me to re-base the dependencies. 

I hadn't tried using requirejs, which may have done the trick. 

thanks! 
Chris 

----- Original Message -----

> Closed #1485 .
> —
> Reply to this email directly or view it on GitHub .

-- 

The Molecule 
594 Broadway Suite 906 
New York, NY 10012 
917.977.1122 (m) 
http://www.themolecule.com
